Create custom systemd-networkd-wait-online.service override to wait on individual interfaces. (LP: #2060311)

slyon opened the pull request:
## Description
Skip s-n-wait-online if we don't have any non-optional interfaces, using a "ConditionPathIsSymbolicLink=" checking Netplan's s-n-wait-online.service enablement symlink.

This is in favor to RequiredForOnline=yes as the behavior of upstream (pure)
systemd-networkd-wait-online.service is not mean to be used in this way.
    
If "RequiredForOnline=no" sd-networkd-wait-online will fully ignore the
corresponding interface and it will block/delay network-online.target if
no interfaces are "RequiredForOnline=yes" at all.
    
FR-7246

Note: This is a replacement for https://github.com/canonical/netplan/pull/455 that keeps compatibility with cloud-init, so it can still sort `After=systemd-networkd-wait-online.service` AND `Before=network-online.target`.

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Adds example YAML for new feature.
- [x] \(Optional\) Closes an open bug in Launchpad. LP#2060311

slyon commented:
PPA for testing can be found in: https://launchpad.net/~slyon/+archive/ubuntu/lp2060311/+packages

slyon commented on src/networkd.c:
```diff
@@ -1499,6 +1499,15 @@ _netplan_networkd_write_wait_online(const NetplanState* np_state, const char* ro
+        // [Install] section via "Also=systemd-networkd-wait-online.service".
+        // We want to avoid it being called without --interface arguments, to
+        // avoid a boot delay, when waiting for configured, but optional interfaces.
+        GString* content_noop = g_string_new("[Service]\nExecStart=\n"
+                                             "ExecStart=/bin/true\n");
```
note: I'm not so sure we want this behavior by default, at least not on non-Ubuntu systems.

daniloegea commented on src/networkd.c:
```diff
@@ -1466,6 +1457,60 @@ _netplan_netdef_write_networkd(
+    // create run/systemd/system/systemd-networkd-wait-online.service.d/
+    const char* override = "/run/systemd/system/systemd-networkd-wait-online.service.d/10-netplan.conf";
+    // The "ConditionPathIsSymbolicLink" is Netplan's s-n-wait-online enablement symlink,
+    // as we want to run -wait-only only if enabled by Netplan.
```
s/wait-only/wait-online/ ?

slyon commented on src/generate.c:
```diff
@@ -300,10 +302,11 @@ int main(int argc, char** argv)
     if (netplan_state_get_backend(np_state) == NETPLAN_BACKEND_NM || any_nm)
         _netplan_g_string_free_to_file(g_string_new(NULL), rootdir, "/run/NetworkManager/conf.d/10-globally-managed-devices.conf", NULL);
 
+    gboolean enable_wait_online = _netplan_networkd_write_wait_online(np_state, rootdir);
```
Note: we should consider writing the s-n-wait-online override only if we have "any_networkd", to avoid fiddling with networkd when nothing is configured through Netplan.

daniloegea commented on src/networkd.c:
```diff
@@ -1466,6 +1457,60 @@ _netplan_netdef_write_networkd(
+    // We have non-optional interface, so let's wait for those explicitly
+    GHashTableIter idx;
+    gpointer ifname;
+    g_string_append(content, "\n[Service]\nExecStart=\n"
```
Why do we need this empty `ExecStart=`?

enr0n replied:
It is valid to have multiple `ExecStart=`. Using an empty one clears the current list so that you can completely override it.

daniloegea commented:
This seem to work well. The new `10-netplan.conf` file can get overcrowded if one has lots of interfaces on Netplan but I guess it shouldn't be a problem (unless you have enough interfaces to reach the maximum number of command line arguments o.O).

I did find a case though where the behavior changes: when you have the loopback interface in your yaml. Apparently `-wait-online -i lo` will always get stuck. So, people with `ethernets.lo` in their configuration will notice that wait-online will start to block after this chance. Maybe `lo` should be a special case and never included in the list?

enr0n commented:
> This seem to work well. The new `10-netplan.conf` file can get overcrowded if one has lots of interfaces on Netplan but I guess it shouldn't be a problem (unless you have enough interfaces to reach the maximum number of command line arguments o.O).
> 
> I did find a case though where the behavior changes: when you have the loopback interface in your yaml. Apparently `-wait-online -i lo` will always get stuck. So, people with `ethernets.lo` in their configuration will notice that wait-online will start to block after this chance. Maybe `lo` should be a special case and never included in the list?

Yeah either ignore it or would have to set a different operational state range, e.g. `-o carrier`. But the default behavior of  systemd-networkd-wait-online (i.e. when no `-i` are specified on the command line) is to ignore loopback.

enr0n reviewed (approved):
The new behavior matches what we have discussed. I did some testing with various combinations of `optional: ` for `eth0` and `wlan0` and things work as expected in all cases.

TheRealFalcon commented:
While I'm not seeing issues specifically related to cloud-init, this doesn't seem to be working correctly, or I'm not understanding how it is supposed to work. After install the PPA, if I reboot, regardless what I have under `/etc/netplan`,  `/run/systemd/system/systemd-networkd-wait-online.service.d/10-netplan.conf` contains:
```
[Unit]
ConditionPathIsSymbolicLink=/run/systemd/generator/network-online.target.wants/systemd-networkd-wait-online.service
```

and nothing else.

slyon commented:
> regardless what I have under `/etc/netplan`, `/run/systemd/system/systemd-networkd-wait-online.service.d/10-netplan.conf` contains:
> 
> ```
> [Unit]
> ConditionPathIsSymbolicLink=/run/systemd/generator/network-online.target.wants/systemd-networkd-wait-online.service
> ```
> 
> and nothing else.

As long as you have any network definition in /etc/netplan/ that uses networkd as a renderer and does NOT use "optional: true" you should see the corresponding interface listed in that file after calling "netplan apply" (or reboot).

TheRealFalcon commented:
Does it matter that I'm in an LXD container?

```
root@asdf:~# cat /etc/netplan/50-cloud-init.yaml 
# This file is generated from information provided by the datasource.  Changes
# to it will not persist across an instance reboot.  To disable cloud-init's
# network configuration capabilities, write a file
# /etc/cloud/cloud.cfg.d/99-disable-network-config.cfg with the following:
# network: {config: disabled}
network:
    renderer: networkd
    ethernets:
        eth0:
            dhcp4: true
            match:
                macaddress: 12:34:56:78:90:ab
            # optional: true
        eth1:
            dhcp4: true
            match:
                macaddress: 00:16:3e:38:c3:9e
            # optional: false
    version: 2
root@asdf:~# netplan apply
root@asdf:~# cat /run/systemd/system/systemd-networkd-wait-online.service.d/10-netplan.conf 
[Unit]
ConditionPathIsSymbolicLink=/run/systemd/generator/network-online.target.wants/systemd-networkd-wait-online.service
root@asdf:~# reboot
root@asdf:~# 
Session terminated, killing shell...
 ...killed.
(cloud-init37) cow:net $ lxc shell asdf
root@asdf:~# cat /run/systemd/system/systemd-networkd-wait-online.service.d/10-netplan.conf 
[Unit]
ConditionPathIsSymbolicLink=/run/systemd/generator/network-online.target.wants/systemd-networkd-wait-online.service
root@asdf:~# 
```

slyon commented:
> Does it matter that I'm in an LXD container?

It worked in an LXD container for me earlier today. But it could have some impact on the macaddress matching.. can you try with a different matching condition, like name or driver? Looks like it doesn't find the interface here. I need to double-check that.

TheRealFalcon commented:
@slyon , doh, sorry. This was after I tore down a previous container and forgot to update my network configuration. Once I corrected the MAC address, it worked for me.

TheRealFalcon commented:
Looks like cloud-init is still having issues first boot. During our init-local timeframe, we write out netplan config, then call `netplan generate`. After boot is finished, checking the `systemd-networkd-wait-online.service`, systemd says we need to call `daemon-reload`.

This can be simulated in a container by running `cloud-init clean --logs; rm /etc/netplan/50-cloud-init.yaml`, then reboot. 

After boot:
```
root@asdf:~# ls /run/systemd/generator/
getty.target.wants  local-fs.target.wants  netplan.stamp
root@asdf:~# systemctl status systemd-networkd-wait-online.service 
Warning: The unit file, source configuration file or drop-ins of systemd-networkd-wait-online.service changed on disk. Run 'systemctl daemon-reload' to reload units.
○ systemd-networkd-wait-online.service - Wait for Network to be Configured
     Loaded: loaded (/usr/lib/systemd/system/systemd-networkd-wait-online.service; enabled; preset: enabled)
    Drop-In: /run/systemd/system/systemd-networkd-wait-online.service.d
             └─10-netplan.conf
     Active: inactive (dead)
  Condition: start condition unmet at Tue 2024-04-16 21:40:25 UTC; 54s ago
             └─ ConditionPathIsSymbolicLink=/run/systemd/generator/network-online.target.wants/systemd-networkd-wait-online.service was not met
       Docs: man:systemd-networkd-wait-online.service(8)

Apr 16 21:40:25 asdf systemd[1]: systemd-networkd-wait-online.service - Wait for Network to be Configured was skipped because of an unmet condition check (ConditionPathIsSymbolicLink=/run/systemd/generator/network-online.target.wants/systemd-networkd-wait>
root@asdf:~# systemctl daemon-reload
root@asdf:~# ls /run/systemd/generator/
getty.target.wants  local-fs.target.wants  multi-user.target.wants  netplan.stamp  network-online.target.wants
root@asdf:~# 
```

daniloegea commented:
Another thing I've found is that `-wait-online` blocks if the interface doesn't exist. So if you have interfaces in your YAMLs that are not present in the system it will block now. The previous behavior is to not block.

slyon commented:
Thanks for all of your testing and comments!

I put all fixes into follow-up commits, so it can be reviewed more easily. @daniloegea PITAL at the most recent commits.
I've built the current version in my PPA as `~ppa6`, and it is fixing:

* no blocking on loopback interface
* no blocking on non-existing interfaces
* no need for manual `daemon-reload` in cloud-init "init-local" timeframe (that is now covered by the `netplan generate` CLI, which is being called anyway)
* Interfaces with IP configuration (dhcp4, dhcp6, static IP or link-local IP) will be listed as `-i IFACE:degraded`
  * Interfaces without IP configuration will be listed as `-i IFACE:carrier`, to avoid blocking on non-existing IPs
* Fixed typos
* no fiddling with `systemd-networkd-wait-online.service.d/10-netplan.conf` override when no networkd-based NetDefs are defined in Netplan's configuration.
  * We do not enable `systemd-networkd.service` in this case, so we should not modify `s-n-wait-online.service` either.
* Included integrationtest/autopkgtest, checking for MAC address matching, too, on a real system/VM
  * see `ethernets.py -> test_systemd_networkd_wait_online`

@TheRealFalcon I addressed the `daemon-reload` failure, and your reproducer seems to be fine now:
```
root@well-pug:~# cloud-init clean --logs; rm /etc/netplan/50-cloud-init.yaml
root@well-pug:~# reboot

$ lxc shell well-pug 

root@well-pug:~# ls /run/systemd/generator/
getty.target.wants  local-fs.target.wants  multi-user.target.wants  netplan.stamp  network-online.target.wants
root@well-pug:~# systemctl status systemd-networkd-wait-online.service 
● systemd-networkd-wait-online.service - Wait for Network to be Configured
     Loaded: loaded (/usr/lib/systemd/system/systemd-networkd-wait-online.service; enabled; preset: enabled)
    Drop-In: /run/systemd/system/systemd-networkd-wait-online.service.d
             └─10-netplan.conf
     Active: active (exited) since Wed 2024-04-17 13:05:40 UTC; 18s ago
       Docs: man:systemd-networkd-wait-online.service(8)
    Process: 301 ExecStart=/lib/systemd/systemd-networkd-wait-online -i eth0:degraded (code=exited, status=0/SUCCESS)
   Main PID: 301 (code=exited, status=0/SUCCESS)
        CPU: 6ms

Apr 17 13:05:39 well-pug systemd[1]: Starting systemd-networkd-wait-online.service - Wait for Network to be Configured.
Apr 17 13:05:40 well-pug systemd[1]: Finished systemd-networkd-wait-online.service - Wait for Network to be Configured.
```

daniloegea commented on src/networkd.c:
```diff
@@ -1460,29 +1474,41 @@ _netplan_netdef_write_networkd(
-                g_hash_table_add(non_optional_interfaces, g_strdup(def->set_name));
+            // Check if we have any IP configuration
+            struct address_iter* addr_iter = _netplan_netdef_new_address_iter(def);
+            gboolean any_ips = _netplan_address_iter_next(addr_iter) != NULL || netplan_netdef_get_link_local_ipv4(def) || netplan_netdef_get_link_local_ipv6(def);
```
question: should dhcp4 and dhcp6 be considered here? When I have the config below I get `enp3s0:carrier` but according to your last comment it should be `:degraded`?

```yaml
network:
  ethernets:
    enp3s0:
      dhcp4: true
      link-local: []
```

slyon replied:
Good catch! Of course, it needs to be considered. Fixing...

daniloegea reviewed (commented):
I was doing some testing based on you comments and I noticed that dhcp is not considered to set the interface as degraded. So if I disabled link-local and enable dhcp4 it will be set to carrier. Is that expected?

blackboxsw commented on src/networkd.c:
```diff
@@ -1460,29 +1474,41 @@ _netplan_netdef_write_networkd(
 {
+    // Set of all current network interfaces, potentially non yet renamed
+    GHashTable* system_interfaces = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    _netplan_query_system_interfaces(system_interfaces);
```
I'm not certain here about the impact of ignoring devices not currently present in /sys/class/net during systemd generator timeframe.  By only considering interfaces already present in /sys/class/net, are we exposed to scenarios where the device drivers for the specified device haven't loaded yet (in initramfs-less environments without the proper drivers compiled into the kernel)?   This reminds me of this cloud-init issue https://github.com/canonical/cloud-init/issues/4451 for minimal images.   In the event that we don't have device drivers loaded (from initramfs or kernel static compiled), is it possible netplan's generator may not see the device in /sys/class/net via `if_nameindex` and systemd-networkd-wait-online will be disabled right?  Maybe in this case, we prescribe that initramfs/kernel must have the desired device drivers in order to expect systemd-networkd-wait-online to block properly.

daniloegea replied:
A similar (the same?) issue will happen with virtual devices. By the time the generator runs they will not be created yet so they will not be in the list.

blackboxsw replied:
James corrected me, this isn't systemd generator timeframe, but the point in time when `netplan  generate` is called.   For cloud-init init-local  boot stage, I think we may still be exposed to this race where cloud-init local gets in any invokes `netplan generate` possibly before supplemental device drivers have loaded. So this may still be a concern.

slyon replied:
Oh! That's a very interesting case.. I just pushed more commits to not require virtual devices (bridges, bonds, dummys, ...) to be available in `/sys/class/net` just yet. Those will be created later.

For physical devices that are not yet loaded, I don't know. I don't think there's anything that can be done as part of this PR and such interfaces might be ignored by `systemd-networkd-wait-online`.

A longer-term solution might be splitting up Netplan into a sytemd-generator and a systemd-service (running `Before=network-pre.target`, similar to `systemd-network-generator.service`), which we are tracking as spec "FO165 - Netplan generator architecture".

Do you think this should be a blocker?

blackboxsw replied:
I think this retains the same gap/issue cloud-init and systemd-networkd-wait-online are already exposed to, so I don't think this needs to be fixed in this PR.  The stance we've taken already [Ubuntu minimal case](https://bugs.launchpad.net/cloud-images/+bug/2036968) was to recompile the kernel to include the module built-ins that are required to support virtio-net to ensure they are loaded in time for network generation and network-online for "initrdless" boots. It feels "ok" to state that if you have a critical device driver needed in early boot for your cloud image and you are booting without initramrd, you would either need to statically compile the necessary device drivers into that kernel, or use initrd and provide any supplemental drivers needed earlier boot. This may just be something we take into consideration for future behavior though, and thinking of how to approach this type of concern later in F0165 probably makes the most sense at the moment.

slyon commented:
Added more commits fixing:
* considering `dhcp4: true` and `dhcp6: true` for "degraded" operational state.
* do not require virtual devices to be created already

daniloegea commented:
Thanks, Lukas. The cases I've found seem to be fixed now.

slyon commented:
Added a tiny fix to recognize that we do not configure link-local IPs on bridge/bond members.
